# Day 14: Extended Polymerization

## Part one

The incredible pressures at this depth are starting to put a strain on your submarine. The submarine has [polymerization](https://en.wikipedia.org/wiki/Polymerization) equipment that would produce suitable materials to reinforce the submarine, and the nearby volcanically-active caves should even have the necessary input elements in sufficient quantities.

The submarine manual contains instructions for finding the optimal polymer formula; specifically, it offers a **polymer template** and a list of **pair insertion** rules (your puzzle input). You just need to work out what polymer would result after repeating the pair insertion process a few times.

For example:

```
NNCB

CH -> B
HH -> N
CB -> H
NH -> C
HB -> C
HC -> B
HN -> C
NN -> C
BH -> H
NC -> B
NB -> B
BN -> B
BB -> N
BC -> B
CC -> N
CN -> C
```

The first line is the **polymer template** - this is the starting point of the process.

The following section defines the **pair insertion** rules. A rule like AB -> C means that when elements A and B are immediately adjacent, element C should be inserted between them. These insertions all happen simultaneously.

So, starting with the polymer template NNCB, the first step simultaneously considers all three pairs:

- The first pair (`NN`) matches the rule `NN -> C`, so element **`C`** is inserted between the first `N` and the second `N`.
- The second pair (`NC`) matches the rule `NC -> B`, so element **`B`** is inserted between the `N` and the `C`.
- The third pair (`CB`) matches the rule `CB -> H`, so element **`H`** is inserted between the `C` and the `B`.

Note that these pairs overlap: the second element of one pair is the first element of the next pair. Also, because all pairs are considered simultaneously, inserted elements are not considered to be part of a pair until the next step.

After the first step of this process, the polymer becomes `NCNBCHB`.

Here are the results of a few steps using the above rules:

```
Template:     NNCB
After step 1: NCNBCHB
After step 2: NBCCNBBBCBHCB
After step 3: NBBBCNCCNBBNBNBBCHBHHBCHB
After step 4: NBBNBNBBCCNBCNCCNBBNBBNBBBNBBNBBCBHCBHHNHCBBCBHCB
```

This polymer grows quickly. After step 5, it has length 97; After step 10, it has length 3073. After step 10, B occurs 1749 times, C occurs 298 times, H occurs 161 times, and N occurs 865 times; taking the quantity of the most common element (B, 1749) and subtracting the quantity of the least common element (H, 161) produces `1749 - 161 =`**`1588`**.

Apply 10 steps of pair insertion to the polymer template and find the most and least common elements in the result. **What do you get if you take the quantity of the most common element and subtract the quantity of the least common element?**

In [1]:
def load_input(filepath='./data/input_test.txt'):
    polymer = ''
    insertion_rules = {}
    with open('%s' % (filepath, ), 'r') as input_f:
        input_lines = input_f.read().splitlines()
        
        # Extract polymer
        polymer = input_lines[0]
        
        # Extract insertion rules
        for line_number in range(2, len(input_lines)):
            input_line = input_lines[line_number]
            line_number += 1
            if not input_line:
                break
            
            pair, element = input_line.split(' -> ')
            insertion_rules[pair] = element
    
    return polymer, insertion_rules

polymer, insertion_rules = load_input()
print('polymer: %s' % (polymer, ))
print('insertion_rules: %s' % (insertion_rules, ))

polymer: NNCB
insertion_rules: {'CH': 'B', 'HH': 'N', 'CB': 'H', 'NH': 'C', 'HB': 'C', 'HC': 'B', 'HN': 'C', 'NN': 'C', 'BH': 'H', 'NC': 'B', 'NB': 'B', 'BN': 'B', 'BB': 'N', 'BC': 'B', 'CC': 'N', 'CN': 'C'}


In [2]:
def insertion(polymer, insertion_rules):
    new_polymer = ''
    for position in range(0, len(polymer) - 1):
        current_char = polymer[position]
        next_char = polymer[position + 1]
        new_polymer += current_char + insertion_rules[current_char + next_char]
        
    new_polymer += polymer[-1]
        
    return new_polymer

def count_elements(polymer):
    count = {}
    
    for position in range(0, len(polymer)):
        element = polymer[position]
        
        if element in count.keys():
            count[element] += 1
        else:
            count[element] = 1
            
    return count

polymer, insertion_rules = load_input()
polymer = insertion(polymer, insertion_rules)
polymer = insertion(polymer, insertion_rules)
polymer = insertion(polymer, insertion_rules)
polymer = insertion(polymer, insertion_rules)
print('polymer: %s' % (polymer, ))
count_elements(polymer)

polymer: NBBNBNBBCCNBCNCCNBBNBBNBBBNBBNBBCBHCBHHNHCBBCBHCB


{'N': 11, 'B': 23, 'C': 10, 'H': 5}

In [3]:
polymer, insertion_rules = load_input('./data/input.txt')

for i in range(0, 10):
    polymer = insertion(polymer, insertion_rules)
    
print('len polymer: %s' % (len(polymer), ))
count = count_elements(polymer)
print('count: %s' % (count, ))
count['O'] - count['C']

len polymer: 19457
count: {'P': 2617, 'H': 1597, 'C': 906, 'O': 4461, 'N': 1239, 'V': 2244, 'B': 1811, 'S': 1645, 'K': 946, 'F': 1991}


3555

Your puzzle answer was `3555`.

# Part two

The resulting polymer isn't nearly strong enough to reinforce the submarine. You'll need to run more steps of the pair insertion process; a total of **40 steps** should do it.

In the above example, the most common element is B (occurring `2192039569602` times) and the least common element is H (occurring `3849876073` times); subtracting these produces **`2188189693529`**.

Apply **40** steps of pair insertion to the polymer template and find the most and least common elements in the result. **What do you get if you take the quantity of the most common element and subtract the quantity of the least common element?**

In [4]:
def extract_element_pairs_ex2(polymer):
    pairs = {}
    
    for position in range(0, len(polymer) - 1):
        pair = polymer[position] + polymer[position + 1]
        
        if pair in pairs.keys():
            pairs[pair] += 1
        else:
            pairs[pair] = 1
    
    return pairs

def insertion_ex2(element_pairs, insertion_rules):
    new_pairs = {}
    
    for key, value in element_pairs.items():
        new_element = insertion_rules[key]
        
        new_pair_1 = key[0] + new_element
        if new_pair_1 in new_pairs.keys():
            new_pairs[new_pair_1] += value
        else:
            new_pairs[new_pair_1] = value
        
        new_pair_2 = new_element + key[1]
        if new_pair_2 in new_pairs.keys():
            new_pairs[new_pair_2] += value
        else:
            new_pairs[new_pair_2] = value
        
    return new_pairs

def count_elements_ex2(element_pairs, last_element):
    count = {}
    count[last_element] = 1
    
    for key, value in element_pairs.items():
        element_1 = key[0]
        element_2 = key[1]
        
        if element_1 in count.keys():
            count[element_1] += value
        else:
            count[element_1] = value

    return count

polymer, insertion_rules = load_input()
element_pairs = extract_element_pairs_ex2(polymer)
for i in range(0, 10):
    element_pairs = insertion_ex2(element_pairs, insertion_rules)

print('element_pairs: %s' % (element_pairs, ))
count_elements_ex2(element_pairs, polymer[-1])

element_pairs: {'NB': 796, 'BB': 812, 'BN': 735, 'BC': 120, 'CC': 60, 'CN': 102, 'NC': 42, 'CB': 115, 'BH': 81, 'HC': 76, 'HH': 32, 'HN': 27, 'NH': 27, 'CH': 21, 'HB': 26}


{'B': 1749, 'N': 865, 'C': 298, 'H': 161}

In [5]:
polymer, insertion_rules = load_input('./data/input.txt')
element_pairs = extract_element_pairs_ex2(polymer)
for i in range(0, 40):
    element_pairs = insertion_ex2(element_pairs, insertion_rules)

count = count_elements_ex2(element_pairs, polymer[-1])
count

{'O': 5195084704903,
 'P': 3178628005225,
 'H': 1378718959755,
 'C': 755642661164,
 'N': 1322198384438,
 'V': 2501671713686,
 'B': 1750740691551,
 'S': 1866024773848,
 'K': 927295863022,
 'F': 2014715170153}

In [6]:
count['O'] - count['C']

4439442043739

Your puzzle answer was `4439442043739`.